# Aviation - Part 1

In [1]:
import numpy as np
import pandas as pd

## 1. Reading the data file

In [2]:
columns  = ['Event Id', 'accident', 'Accident Number', 'Event Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport Code',
       'Airport Name', 'Injury Severity', 'Aircraft Damage',
       'Aircraft Category', 'Registration Number', 'Make', 'Model',
       'Amateur Built', 'Number of Engines', 'Engine Type',
       'FAR Description', 'Schedule', 'Purpose of Flight', 'Air Carrier',
       'Total Fatal Injuries', 'Total Serious Injuries',
       'Total Minor Injuries', 'Total Uninjured', 'Weather Condition',
       'Broad Phase of Flight', 'Report Status', 'Publication Date',
       'todel1']

In [3]:
parse_dates=["Event Date"]

In [4]:
converters= { 'accident': lambda x: 1 if x.strip()=='Accident' else 0
             ,'Country': lambda x: x.strip()
             ,'Aircraft Category': lambda x: x.strip()
             ,'Location': lambda x: x.strip()
            }

In [5]:
df = pd.read_csv('AviationData.txt', sep='|'
                 , header=0, names=columns
                 , converters=converters
                 , parse_dates=parse_dates)

## size of the dataset
df.shape

(77592, 32)

### 1.1 Filtering

In [6]:
#Filter by country: "United States"
country = "United States"
df = df[df.Country == country]

In [7]:
#Filter by Aircraft Category: "Airplane"
aircraft_category = "Airplane"
df = df[df['Aircraft Category'] == aircraft_category]

### 1.2 Creating the _state_ column

In [8]:
# splitting

#extract state from location 
df['state'] = df['Location'].map(lambda x: x.split(',')[-1].strip())

### 1.3 More filtering

In [9]:
## filtering empty state rows
df = df[df['state'] != '']

## 2. Some Statistics

In [10]:
df['state'].value_counts().to_json()

'{"CA":1280,"TX":1011,"FL":887,"AK":875,"AZ":486,"CO":425,"WA":405,"GA":372,"IL":332,"NY":325,"NC":298,"MI":286,"OH":284,"ID":279,"PA":266,"OR":265,"MO":261,"WI":239,"MN":230,"AR":229,"VA":218,"NM":210,"IN":205,"KS":204,"UT":202,"NV":197,"OK":192,"LA":191,"NJ":184,"AL":182,"MT":179,"SC":179,"TN":177,"IA":159,"MD":144,"WY":135,"MS":134,"MA":129,"KY":126,"NE":121,"ME":93,"ND":86,"SD":76,"CT":69,"WV":63,"HI":50,"NH":49,"PR":33,"VT":32,"RI":22,"DE":16,"GU":8,"DC":7,"VI":7,"GM":5,"AO":3,"PO":3,"FN":2,"MP":2,"ON":1,"AS":1,"UN":1}'

In [11]:
len(df['state'].value_counts().keys())

62

#### Observation

But USA is a federal republic consisting of 50 states, one federal district (Washington, D.C.) and 5 inhabited territories (American Samoa, Guam, the Northern Mariana Islands, Puerto Rico, and the U.S. Virgin Islands)  making a total of 56 posible states but in dataset are 62 states. Some of this additional 'states' are: Atlantic Ocean (AO), Gulf of Mexic(GM), Pacific Ocean(PO), etc.

In [12]:
df['state'].value_counts().keys()

Index(['CA', 'TX', 'FL', 'AK', 'AZ', 'CO', 'WA', 'GA', 'IL', 'NY', 'NC', 'MI',
       'OH', 'ID', 'PA', 'OR', 'MO', 'WI', 'MN', 'AR', 'VA', 'NM', 'IN', 'KS',
       'UT', 'NV', 'OK', 'LA', 'NJ', 'AL', 'MT', 'SC', 'TN', 'IA', 'MD', 'WY',
       'MS', 'MA', 'KY', 'NE', 'ME', 'ND', 'SD', 'CT', 'WV', 'HI', 'NH', 'PR',
       'VT', 'RI', 'DE', 'GU', 'DC', 'VI', 'GM', 'AO', 'PO', 'FN', 'MP', 'ON',
       'AS', 'UN'],
      dtype='object')

In [13]:
#using a dictionary of states from: http://code.activestate.com/recipes/577305-python-dictionary-of-us-states-and-territories/
states = {
        'AK': 'Alaska',
        'AL': 'Alabama',
        'AR': 'Arkansas',
        'AS': 'American Samoa',
        'AZ': 'Arizona',
        'CA': 'California',
        'CO': 'Colorado',
        'CT': 'Connecticut',
        'DC': 'District of Columbia',
        'DE': 'Delaware',
        'FL': 'Florida',
        'GA': 'Georgia',
        'GU': 'Guam',
        'HI': 'Hawaii',
        'IA': 'Iowa',
        'ID': 'Idaho',
        'IL': 'Illinois',
        'IN': 'Indiana',
        'KS': 'Kansas',
        'KY': 'Kentucky',
        'LA': 'Louisiana',
        'MA': 'Massachusetts',
        'MD': 'Maryland',
        'ME': 'Maine',
        'MI': 'Michigan',
        'MN': 'Minnesota',
        'MO': 'Missouri',
        'MP': 'Northern Mariana Islands',
        'MS': 'Mississippi',
        'MT': 'Montana',
        'NA': 'National',
        'NC': 'North Carolina',
        'ND': 'North Dakota',
        'NE': 'Nebraska',
        'NH': 'New Hampshire',
        'NJ': 'New Jersey',
        'NM': 'New Mexico',
        'NV': 'Nevada',
        'NY': 'New York',
        'OH': 'Ohio',
        'OK': 'Oklahoma',
        'OR': 'Oregon',
        'PA': 'Pennsylvania',
        'PR': 'Puerto Rico',
        'RI': 'Rhode Island',
        'SC': 'South Carolina',
        'SD': 'South Dakota',
        'TN': 'Tennessee',
        'TX': 'Texas',
        'UT': 'Utah',
        'VA': 'Virginia',
        'VI': 'Virgin Islands',
        'VT': 'Vermont',
        'WA': 'Washington',
        'WI': 'Wisconsin',
        'WV': 'West Virginia',
        'WY': 'Wyoming'
}

__exploring unknown states__

In [14]:
unknown_states = []
for k in df['state'].value_counts().keys():
    if not k in states:
        unknown_states.append(k)
print(unknown_states)


['GM', 'AO', 'PO', 'FN', 'ON', 'UN']


In [15]:
df.Location[df.state.isin(unknown_states)]

467             Leixlip, FN
1139          Barcelona, FN
1187           HONOLULU, PO
1446            Whitney, ON
2627            Seattle, PO
4223      Pacific Ocean, PO
4790     Gulf of Mexico, GM
5428     Atlantic Ocean, AO
5521     Atlantic Ocean, AO
6698            Montauk, AO
6975            Unknown, GM
8216     Gulf of Mexico, GM
9980     Gulf of Mexico, GM
10821    Gulf of Mexico, GM
77589           Unknown, UN
Name: Location, dtype: object

In [16]:
df[df.state.isin(unknown_states)]

Event Id  accident Accident Number Event Date  \
467    20150812X90823          0    CEN15WA358B  2015-07-25   
1139   20150407X75254          1     CEN15WA192  2015-03-04   
1187   20150219X14806          1     DCA15LA067  2015-02-15   
1446   20141120X02000          1     CEN15WA052  2014-11-11   
2627   20140320X53321          1     DCA14CA083  2014-03-17   
4223   20130220X51849          1     WPR13LA131  2013-02-19   
4790   20120921X52540          1     CEN12LA652  2012-09-20   
5428   20120623X90739          1     ERA12LA410  2012-06-22   
5521   20120616X31702          1     ERA12FAMS2  2012-06-08   
6698   20111013X11758          1     DCA11WA110  2011-09-12   
6975   20110804X94823          1     ERA11FAMS1  2011-08-01   
8216   20101207X95916          1     CEN11LA097  2010-12-04   
9980   20091222X33027          1     ERA10FAMS1  2009-12-15   
10821  20090709X74830          1     ERA09FA389  2009-07-08   
77589  20130128X92153          1     WPR12TA445  2015-11-29   

                 Location        Country     Latitude      Longitude  \
467           Leixlip, FN  United States                               
1139        Barcelona, FN  United States                               
1187         HONOLULU, PO  United States   22.608889    -153.843055    
1446          Whitney, ON  United States   45.377223     -78.298055    
2627          Seattle, PO  United States    0.000000       0.000000    
4223    Pacific Ocean, PO  United States   20.228056    -139.203056    
4790   Gulf of Mexico, GM  United States                               
5428   Atlantic Ocean, AO  United States   42.454444     -69.914444    
5521   Atlantic Ocean, AO  United States   39.670000     -72.221389    
6698          Montauk, AO  United States   40.785277     -71.526389    
6975          Unknown, GM  United States   29.079445     -88.631667    
8216   Gulf of Mexico, GM  United States   27.633889     -89.133889    
9980   Gulf of Mexico, GM  United States   27.366111     -85.433611    
10821  Gulf of Mexico, GM  United States   28.445555     -83.122500    
77589         Unknown, UN  United States                               

      Airport Code                      Airport Name  ...   \
467                                                   ...    
1139                                                  ...    
1187                                            N/A   ...    
1446                                            N/A   ...    
2627                                            N/A   ...    
4223          SFO    San Francisco International Ai   ...    
4790                                                  ...    
5428                                           None   ...    
5521                                                  ...    
6698                                                  ...    
6975                                                  ...    
8216                                                  ...    
9980          SPG                    Albert Whitted   ...    
10821                                                 ...    
77589                                                 ...    

      Total Fatal Injuries Total Serious Injuries Total Minor Injuries  \
467                                                                      
1139                                           1                         
1187                                           1                    4    
1446                    2                                                
2627                                           1                         
4223                                           1                         
4790                                                                2    
5428                                                                     
5521                    1                                                
6698                                           2                         
6975                    1                    

__deleting rows with  non-US states__

In [17]:
for x in ['GM', 'PO', 'AO', 'FN', 'UN', 'ON']:
    df = df[df['state'] != x]

__*oldest and newest records*__

In [18]:
df['Event Date'].min(), df['Event Date'].max()

(Timestamp('1982-01-01 00:00:00'), Timestamp('2015-11-29 00:00:00'))

## 3. Saving data for visualization

In [19]:
df.shape

(13117, 33)

__the dataset was reduced from an initial of 77592 rows to 13117__

In [20]:
df['state'].value_counts().to_json('graphs/data01.json')